Bug: test_gnmi_configdb_polling_01 fails on IPv6-only management DUTs due to invalid gnmi_cli -a address after VRF parameterization (#20456)

### Is it platform specific

generic

### Importance or Severity

Critical

### Description of the bug

@spandan-nexthop , After https://github.com/sonic-net/sonic-mgmt/pull/20456, test_gnmi_configdb_polling_01 fails deterministically on DUTs with IPv6-only management addresses.


### Steps to Reproduce

Run test_gnmi_configdb_polling_01 on DUTs with IPv6-only management addresses

### Actual Behavior and Expected Behavior

Actual Behavior:
gnmi_subscribe_polling() builds an invalid -a <ipv6-address>:<port> target because IPv6 bracket formatting is skipped when ip is passed explicitly, causing gnmi_cli to time out and return empty output. The test then fails because msg.count("bgp_asn") is 0 instead of the expected 3.

Expected Behavior:
gnmi_subscribe_polling() should format IPv6 addresses as [<ipv6-address>]:<port> regardless of whether ip is defaulted or explicitly provided. gnmi_cli should connect successfully and return polling responses containing bgp_asn at least 3 times.

### Relevant log output

```shell
AssertionError: Subscribe table key for DEVICE_METADATA: 
]
duthosts = []
rand_one_dut_hostname = 'r-panther-02'
ptfhost = 
test_data = {'name': 'Subscribe table key for DEVICE_METADATA', 'path': '/sonic-db:CONFIG_DB/localhost/DEVICE_METADATA/localhost'}
vrf_config = {'description': 'Default (no VRF)', 'dut_intf': 'eth0', 'dut_ip': '2001:db8:1234:5678:a288:c2ff:fe86:5c40', 'name': 'default_1', ...}

    @pytest.mark.parametrize('test_data', test_data_metadata)
    def test_gnmi_configdb_polling_01(duthosts, rand_one_dut_hostname, ptfhost, test_data, vrf_config):
        '''
        Verify GNMI subscribe API, streaming onchange mode
        Subscribe polling mode
        '''
        duthost = duthosts[rand_one_dut_hostname]
        exp_cnt = 3
        path_list = [test_data["path"]]
        msg, _ = gnmi_subscribe_polling(
            duthost, ptfhost, path_list, 1000, exp_cnt, ip=vrf_config["dut_ip"], vrf_name=vrf_config["vrf"]
        )
>       assert msg.count("bgp_asn") >= exp_cnt, test_data["name"] + ": " + msg
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E       AssertionError: Subscribe table key for DEVICE_METADATA: 
E       ]
```

### Output of `show version`

```shell

```

### Attach files (if any)

_No response_

--- a/tests/gnmi/helper.py
+++ b/tests/gnmi/helper.py
@@ -285,7 +285,7 @@ def gnmi_set(duthost, ptfhost, delete_list, update_list, replace_list, cert=None
     if rc != 0 or "GRPC error" in combined or "rpc error" in combined:
         dump_gnmi_log(duthost)
         dump_system_status(duthost)
-        raise Exception("py_gnmicli failed rc={}\nSTDOUT:\n{}\nSTDERR:\n{}".format(rc, stdout, stderr))
+        raise Exception(f"py_gnmicli failed rc={rc}\nSTDOUT:\n{stdout}\nSTDERR:\n{stderr}")  # noqa: E231
 
 
 def gnmi_get(duthost, ptfhost, path_list, ip=None):

--- a/tests/gnmi/helper.py
+++ b/tests/gnmi/helper.py
@@ -2,6 +2,7 @@ import time
 import logging
 import pytest
 import json
+import ipaddress
 from tests.common.utilities import wait_until
 from tests.common.platform.device_utils import get_dpu_ip, get_dpu_port
 from tests.common.helpers.gnmi_utils import GNMIEnvironment, add_gnmi_client_common_name, del_gnmi_client_common_name, \
@@ -334,6 +335,32 @@ def gnmi_get(duthost, ptfhost, path_list, ip=None):
         raise Exception("error:" + msg)
 
 
+def _bracket_ipv6(ip):
+    """Wrap an IPv6 literal in brackets for use in a host:port target.
+
+    gnmi_cli/gnoi_client take the server address as ``-a <host>:<port>``.
+    A bare IPv6 address contains colons that collide with the ``:<port>``
+    separator, so IPv6 literals must be written as ``[<ipv6>]:<port>``.
+    IPv4 addresses, hostnames and already-bracketed IPv6 values are returned
+    unchanged.
+
+    Args:
+        ip: server address (IPv4/IPv6 literal, hostname, or ``[ipv6]``).
+
+    Returns:
+        The address with IPv6 literals wrapped in brackets.
+    """
+    if not ip or ip.startswith('['):
+        return ip
+    try:
+        if ipaddress.ip_address(ip).version == 6:
+            return f"[{ip}]"
+    except ValueError:
+        # Not an IP literal (e.g. a hostname); leave it untouched.
+        pass
+    return ip
+
+
 # py_gnmicli does not fully support POLLING mode
 # Use gnmi_cli instead
 def gnmi_subscribe_polling(duthost, ptfhost, path_list, interval_ms, count, ip=None, vrf_name=None):
@@ -358,8 +385,13 @@ def gnmi_subscribe_polling(duthost, ptfhost, path_list, interval_ms, count, ip=N
         return "", ""
     env = GNMIEnvironment(duthost, GNMIEnvironment.GNMI_MODE)
     if ip is None:
-        dut_facts = duthost.dut_basic_facts()['ansible_facts']['dut_basic_facts']
-        ip = f"[{duthost.mgmt_ip}]" if dut_facts.get('is_mgmt_ipv6_only', False) else duthost.mgmt_ip
+        ip = duthost.mgmt_ip
+    # gnmi_cli expects the target as -a <host>:<port>. IPv6 literals must be
+    # wrapped in brackets ([<ipv6>]:<port>) so the address colons are not
+    # confused with the host:port separator. Normalize here so bracketing is
+    # applied whether `ip` was defaulted from duthost.mgmt_ip or passed in
+    # explicitly (e.g. a bare IPv6 dut_ip from a VRF config).
+    ip = _bracket_ipv6(ip)
     port = env.gnmi_port
     interval = interval_ms / 1000.0
     # For a non-default VRF the gnmi container does not have `ip vrf exec`

